### Reduction operation: the sum of the numbers in the range [0, a)

In [4]:
import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n")

Time taken by reduction operation: 0.26982951164245605 seconds
256 ms ± 3.2 ms per loop (mean ± std. dev. of 2 runs, 1 loop each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



In [5]:
import time

# Crear una lista con 10**6 elementos
lista = list(range(10**6))

# Suma usando un bucle for
suma_for = 0
start_time = time.time()
for num in lista:
    suma_for += num
end_time = time.time()

print("Suma con bucle for:", suma_for)
print("Tiempo con bucle for:", end_time - start_time, "segundos")

# Suma usando la función sum()
start_time = time.time()
suma_builtin = sum(lista)
end_time = time.time()

print("Suma con sum():", suma_builtin)
print("Tiempo con sum():", end_time - start_time, "segundos")


Suma con bucle for: 499999500000
Tiempo con bucle for: 0.46884942054748535 segundos
Suma con sum(): 499999500000
Tiempo con sum(): 0.030940771102905273 segundos


In [6]:
import numpy as np
import time

# Convertir la lista a un array de NumPy
array = np.array(lista)

# Suma usando un bucle for
suma_for_numpy = 0
start_time = time.time()
for num in array:
    suma_for_numpy += num
end_time = time.time()

print("Suma con bucle for (NumPy):", suma_for_numpy)
print("Tiempo con bucle for (NumPy):", end_time - start_time, "segundos")

# Suma usando np.sum() (vectorizada y optimizada)
start_time = time.time()
suma_numpy = np.sum(array)
end_time = time.time()

print("Suma con np.sum():", suma_numpy)
print("Tiempo con np.sum():", end_time - start_time, "segundos")


Suma con bucle for (NumPy): 499999500000
Tiempo con bucle for (NumPy): 1.5631029605865479 segundos
Suma con np.sum(): 499999500000
Tiempo con np.sum(): 0.004665374755859375 segundos


### Explicación de Resultados

1. **Código Original (Secuencial con `reduc_operation`)**:
    - Este código usa un bucle `for` para calcular la suma directamente sobre el rango `[0, a)`.
    - **Tiempo**: ~256 ms (con `%timeit`).
    - La implementación es sencilla, pero ineficiente porque recorre todos los elementos uno por uno.

2. **Listas en Python**:
    - **Bucle for**: ~0.469 s.
        - Similar al código original en cuanto a lógica, pero aplicando el bucle sobre una lista creada con `range`.
        - Más lento que usar la función `sum` debido a la sobrecarga del bucle explícito.
    - **Función `sum()`**: ~0.031 s.
        - La función `sum()` es mucho más rápida porque está optimizada internamente en C, evitando la iteración explícita.

3. **Arrays en NumPy**:
    - **Bucle for**: ~1.563 s.
        - Sorprendentemente, el bucle explícito en NumPy es aún más lento que en listas de Python.
        - Esto ocurre porque NumPy no está diseñado para operaciones elementales iterativas; su fortaleza está en las operaciones vectorizadas.
    - **`np.sum()`**: ~0.005 s.
        - Este es el método más rápido de todos. NumPy aprovecha operaciones vectorizadas optimizadas en C, lo que reduce significativamente el tiempo de ejecución.

---

### Conclusión

- El uso de **NumPy con operaciones vectorizadas (`np.sum()`)** es significativamente más eficiente, siendo aproximadamente **50 veces más rápido** que el bucle `for` en listas de Python.
- La función `sum()` en listas también es eficiente, pero menos que las optimizaciones de NumPy.
- Los bucles explícitos en general son los métodos más lentos, especialmente cuando se aplican a estructuras como arrays de NumPy, que están diseñados para operaciones vectorizadas.
